src/compiler/checker.ts

```diff
@@ -33219,128 +33181,379 @@ namespace ts {
+            return tryCreateTypeReference(getGlobalClassAccessorDecoratorResultType(/*reportError*/ true), [thisType, inputType, outputType]);
         }
 
-        function createClassAccessorESDecoratorResultType(thisType: Type, valueInType: Type, valueOutType: Type) {
-            const globalClassAccessorDecoratorResultType = getGlobalClassAccessorDecoratorResultType(/*reportError*/ true);
-            const resultType = globalClassAccessorDecoratorResultType !== emptyGenericType ?
-                createTypeReference(globalClassAccessorDecoratorResultType, [thisType, valueInType, valueOutType]) :
-                unknownType;
-            return getUnionType([resultType, voidType]);
+        function createClassFieldDecoratorInitializerMutatorType(thisType: Type, inputType: Type, outputType: Type) {
+            const thisParam = createParameter("this" as __String, thisType);
+            const valueParam = createParameter("value" as __String, outputType);
+            return createFunctionType(/*typeParameters*/ undefined, thisParam, [valueParam], inputType, /*typePredicate*/ undefined, 1);
         }
 
-        function createClassFieldESDecoratorReturnType(valueInType: Type, valueOutType: Type) {
-            const param = createSymbol(SymbolFlags.FunctionScopedVariable, "value" as __String, /*checkFlags*/ undefined, valueOutType);
-            const initializerSignatureType = createFunctionType(/*typeParameters*/ undefined, /*thisParameter*/ undefined, [param], valueInType);
-            return getUnionType([initializerSignatureType, voidType]);
+        /**
+         * Creates a call signature for an ES Decorator. This method is used by the semantics of
+         * `getESDecoratorCallSignature`, which you should probably be using instead.
+         */
+        function createESDecoratorCallSignature(targetType: Type, contextType: Type, nonOptionalReturnType: Type) {
+            const targetParam = createParameter("target" as __String, targetType);
+            const contextParam = createParameter("context" as __String, contextType);
+            const returnType = getUnionType([nonOptionalReturnType, voidType]);
+            return createCallSignature(/*typeParameters*/ undefined, /*thisParameter*/ undefined, [targetParam, contextParam], returnType);
         }
 
+        /**
+         * Gets a call signature that should be used when resolving `decorator` as a call. This does not use the value
+         * of the decorator itself, but instead uses the declaration on which it is placed along with its relative
+         * position amongst other decorators on the same declaration to determine the applicable signature. The
+         * resulting signature can be used for call resolution, inference, and contextual typing.
+         */
         function getESDecoratorCallSignature(decorator: Decorator) {
+            // In the future we are considering allowing the types of decorators to affect the type of the class and its

```

Comment:
'we' here is TS team, not the ES spec committee, right?

Reply:
Yes, the TS team.